### PR TITLE
Add unit tests for canonicalize

### PR DESCRIPTION
**Pull request by boolean5**
This PR does the following and will eventually close https://github.com/mozilla-services/shavar-list-creation/issues/135:

* Adds `pytest` as a dependency and pins it to the [last version that supports Python 2](https://docs.pytest.org/en/latest/py27-py34-deprecation.html) (`v4.6.9`)
* Implements a unit test that validates the canonicalization implementation, using some of the suggested test cases from the [Safe Browsing v2 API documentation](https://web.archive.org/web/20160422212049/https://developers.google.com/safe-browsing/developers_guide_v2#Canonicalization), with some adjustments and additions
* Implements a unit test that confirms that the return type of `canonicalize` is `str`
* Adds instructions for running the unit tests in the `README`

A few notes:

1. Some of the suggested test cases were failing, as I mentioned in https://github.com/mozilla-services/shavar-list-creation/issues/135#issuecomment-634774083. Those have been omitted for now. After we discuss them and fix possible bugs in `canonicalize` I will include them, either in this PR or in a subsequent one.

2. I have added a few test cases in order to test the implementation of some canonicalization rules that were not covered by the suggested test cases in the Safe Browsing docs:
        a. A test case for removing leading dots from hostnames 
        b. A test case for replacing consecutive dots in hostnames with a single dot
        c. A test case that confirms we do not apply path canonicalization rules to query parameters

3. If we later end up with multiple test dependencies as more tests are added, we should probably move them to a separate requirements file.

4. If we want the tests to be less fine-grained, we can also group multiple assertions in multiple test functions according to the concept that they are testing, instead of parametrizing a single test function with all the test cases. For example, `test_canonicalize_unescape_url()` would contain the test cases with `ids` `"unescape_url_1"` through `"unescape_url_7"`, etc.

5. I think some of the suggested test cases in the Safe Browsing docs  are overlapping in the concepts they are testing, so we could reduce the test cases a little if we wanted to.

### CODE REVIEW

File: tests/test_lists2safebrowsing.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+@pytest.mark.parametrize("url,expected",
+                         [pytest.param(url, expected, id=id)
+                             for id, url, expected in CANONICALIZE_TESTCASES])
+def test_canonicalize(url, expected):
+    """Validate the canonicalization implementation.
+
+    Use the test cases suggested in the Safe Browsing v2 API
+    documentation with some adjustments and additions.
+    """
+    assert canonicalize(url) == expected
```
Comment (say-yawn) [inline tests/test_lists2safebrowsing.py:68]:
```suggestion
    assert canonicalize(url) == expected, "Test failed on: %s" % id
```
That way when a test fails we can see which case it failed on.

**Comment by say-yawn** [comment]
> * Some of the suggested test cases were failing, as I mentioned in [#135 (comment)](https://github.com/mozilla-services/shavar-list-creation/issues/135#issuecomment-634774083). Those have been omitted for now. After we discuss them and fix possible bugs in `canonicalize` I will include them, either in this PR or in a subsequent one.
Responded [here](https://github.com/mozilla-services/shavar-list-creation/issues/135#issuecomment-636062326)

> * I have added a few test cases in order to test the implementation of some canonicalization rules that were not covered by the suggested test cases in the Safe Browsing docs:
>   a. A test case for removing leading dots from hostnames
>   b. A test case for replacing consecutive dots in hostnames with a single dot
>   c. A test case that confirms we do not apply path canonicalization rules to query parameters
All these test cases would be nice to have.

> * If we later end up with multiple test dependencies as more tests are added, we should probably move them to a separate requirements file.
Agreed. It's not needed now but I suggest getting a doc or comment in requirements on what is or is not needed for test.

> * If we want the tests to be less fine-grained, we can also group multiple assertions in multiple test functions according to the concept that they are testing, instead of parametrizing a single test function with all the test cases. For example, `test_canonicalize_unescape_url()` would contain the test cases with `ids` `"unescape_url_1"` through `"unescape_url_7"`, etc.
I think separating them out is better since it makes it more clear what it's testing.

> * I think some of the suggested test cases in the Safe Browsing docs  are overlapping in the concepts they are testing, so we could reduce the test cases a little if we wanted to.
Let's keep it as is. There may be overlap but better be more thorough than accidentally remove cases that need to be tested.

File: tests/test_lists2safebrowsing.py (hunk 1 above)
Comment (boolean5) [reply tests/test_lists2safebrowsing.py:68]: Even if `pytest` is called without the verbosity flag, the failure output begins with a line that contains the id of the test (e.g.,  `__________ test_canonicalize[query_parameters_3]
__________`), so we will always know the specific case it failed on. Is there another reason why we need this? I'll be happy to add it if you think it's still needed.

**Comment by say-yawn** [comment]
Woops, this was not supposed to be closed.

**Comment by boolean5** [comment]
I stumbled upon another bug in `canonicalize()`. It only removes consecutive slashes if they are found at the beginning of the path. For instance, `canonicalize("www.example.com//foo//bar")` returns `"www.example.com/foo//bar"` when it should be returning `"www.example.com/foo/bar"`. I've pushed a commit that fixes this.

File: tests/test_lists2safebrowsing.py (hunk 1 above)
Comment (say-yawn) [reply tests/test_lists2safebrowsing.py:99]: I see, it seems I wasn't looking for the information at the right place. Thanks for catching that.

File: tests/test_lists2safebrowsing.py [hunk 2]
```diff
@@ -7,7 +7,9 @@
     ("no_change_1", "www.google.com/", "www.google.com/"),
     ("no_change_2", "evil.com/foo;", "evil.com/foo;"),
     ("remove_scheme_1", "http://www.google.com/", "www.google.com/"),
-    ("remove_scheme_2", "https://www.securesite.com/", "www.securesite.com/"),
+    ("remove_scheme_2", "https://base-fingerprinting-track-digest256."
```
[the remark this answers is not in the document]
Comment (boolean5) [reply tests/test_lists2safebrowsing.py]: That's a good point. Ok, done!

**Pull request merged**